packages/codegen-ui-react/lib/workflow/mutation.ts

```diff
@@ -246,17 +247,21 @@ export function filterStateReferencesForComponent(
 ): MutationReferences {
   return stateReferences
     .filter(({ componentName }) => componentName === component.name)
-    .reduce(mutationReferenceReducer, {});
+    .reduce(mutationReferenceReducerWithComponentType(component.componentType), {});
 }
 
-function mutationReferenceReducer(
-  mutationReferences: MutationReferences,
-  stateReference: StateReference,
-): MutationReferences {
-  const propertyReferences =
-    stateReference.property in mutationReferences ? mutationReferences[stateReference.property] : [];
-  return {
-    ...mutationReferences,
-    [stateReference.property]: propertyReferences.concat([{ addControlEvent: !('set' in stateReference) }]),
+function mutationReferenceReducerWithComponentType(componentType: string) {
+  return function mutationReferenceReducer(
+    mutationReferences: MutationReferences,
+    stateReference: StateReference,
+  ): MutationReferences {
+    const propertyReferences =
+      stateReference.property in mutationReferences ? mutationReferences[stateReference.property] : [];
+    return {
+      ...mutationReferences,
+      [stateReference.property]: propertyReferences.concat([
+        { addControlEvent: PrimitivesWithChangeEvent.has(componentType as Primitive) },
+      ]),
```

Comment:
Wouldn't we want the addControlEvent flag here to effectively be either the old logic or this new logic? Is there a usecase where we want to add a control event to a non-form element? Could be not, but it may be safer in the short term to leave the old behavior in until we convince our selves otherwise.